lib/types.js

```diff
@@ -1950,29 +1950,32 @@ ArrayType.prototype._check = function (val, flags, hook, path) {
 
 ArrayType.prototype._read = function (tap) {
   var items = this.itemsType;
-  var val = [];
+  var val;
   var n;
+  var i = 0
   while ((n = tap.readLong())) {
     if (n < 0) {
       n = -n;
       tap.skipLong(); // Skip size.
     }
+    val = val || new Array(n)
```

Comment:
Have you tried playing the same trick for subsequent batches? Something like:

```js
if (val) {
  val.length += n;
} else {
  val = new Array(n);
}
```

Intuitively it would give similar benefits to the ones this change provides for the first batch.

Reply:
Unfortunately, v8 does not really like `.length` writes. Here is a quick benchmark which demonstrates the fact
https://jsbench.me/g3kmswtm4m/1 and I have seen similar results in other contexts as well

Reply:
Interesting, thanks for the benchmark.